### PR TITLE
Add documentation for the new parent-join field

### PR DESCRIPTION
**Pull request by jimczi**
This commit adds the docs for the new parent-join field.
It explains how to define, index and query this new field.

Relates #20257

### CODE REVIEW

File: docs/reference/mapping/types/parent-join.asciidoc [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+      "properties": {
+        "my_join_field": { <1>
+          "type": "join",
+          "my_parent": "my_child" <2>
+        }
+      }
+    }
+  }
+}
+--------------------------------------------------
```
Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: I think it'd be easier to follow if you had a complete example with a single parent and child relationship and an example with more than one. The way it is now, when I read the document I start to have questions about how to use the field and then I have to scan below the ways to configure multiple parents before I get to the "how".

File: docs/reference/mapping/types/parent-join.asciidoc [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+// CONSOLE
+
+<1> The name for the field
+<2> `my_parent` is parent of `my_child`.
```
Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: I think this bit deserves more explanation.

Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: Maybe it'd be more clear to index a parent with some children. Then it'd be more obvious how `my_parent` and `my_child` are used. It looks like `my_parent` as a key has special meaning and I'm not used to that sort of thing in mappings. We do it all the time in queries though.

File: docs/reference/mapping/types/parent-join.asciidoc (hunk 1 above)
Comment (jimczi) [inline docs/reference/mapping/types/parent-join.asciidoc]: I 
 changed the examples in https://github.com/elastic/elasticsearch/pull/25227/commits/a088fcf0794eeb0432249472b2221350990e86c0 with a focus on single parent/child relation.

File: docs/reference/mapping/types/parent-join.asciidoc [hunk 3]
```diff
@@ -0,0 +1,399 @@
+[[mapping-parent-join-field]]
+=== `join` datatype
+
+The `join` datatype is a special field that creates
+parent/child relation within documents of the same index.
+This field defines a set of possible relations within the documents,
+each relation being a parent name and a child name.
+A parent/child relation can be defined as follow:
```
Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: s/follow/follows/

File: docs/reference/mapping/types/parent-join.asciidoc [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    }
+  }
+}
+--------------------------------------------------
+// CONSOLE
+
+<1> `my_parent` is parent of `my_child` and `another_child`
+<2> `another_child` is parent of `grand_child
+
+Indexing a great child document requires a `routing` value equals
```
Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: s/great/grand/

File: docs/reference/mapping/types/parent-join.asciidoc [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+--------------------------------------------------
+PUT my_index
+{
+  "mappings": {
+    "doc": {
+      "properties": {
+        "my_join_field": {
+          "type": "join",
+          "my_parent": ["my_child", "another_child"],  <1>
+          "another_child": "grand_child" <2>
```
Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: It isn't clear to me from reading the docs what this does. Does it make something like:

```
     my_parent
     /      |
    /       |
my_child  another_child
            |
            |
       grand_child
```
?

Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: Maybe if you are going to build something this complex in a doc you should add the ascii art?

Comment (jimczi) [inline docs/reference/mapping/types/parent-join.asciidoc]: I pushed https://github.com/elastic/elasticsearch/pull/25227/commits/44de7acec5789b0ce9ced16121a5fbd82735f091 with the example tree, I think the render is ok without the ascii art.

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
Thanks for the ascii art!

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
Oh! The docs build didn't quite work for me locally. I think you have some mismatching callouts?

File: docs/reference/mapping/types/parent-join.asciidoc [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+                    "4"
+                ]
+            }
+        ]
+    }
+}
+--------------------------------------------------
+// TESTRESPONSE[s/\.\.\./"timed_out": false, "took": $body.took, "_shards": $body._shards/]
+
+<1> This document belongs to the `my_parent` join
```
Comment (nik9000) [inline docs/reference/mapping/types/parent-join.asciidoc]: I think these are missing a callout in the example above.

**Comment by jimczi** [comment]
Thanks @nik9000 and @martijnvg ! I'll backport to 5.x

**Pull request merged**